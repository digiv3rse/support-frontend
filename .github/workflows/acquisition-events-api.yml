name: Build acquisition-events-api

on:
  pull_request:
    branches:
      - main
  push:
    branches:
      - main

jobs:
  CI:
    if: >-
      (github.actor != 'dependabot[bot]') &&
        (github.event.pull_request.head.repo.owner.login == 'guardian' ||
          github.event_name == 'push')

    # Required by actions-assume-aws-role
    permissions:
      id-token: write
      contents: read

    name: acquisition-events-api build
    runs-on: ubuntu-latest
    steps:
      - name: Env
        run: env

      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node for CDK
        uses: actions/setup-node@v4
        with:
          node-version: 16
          cache: "yarn"
          cache-dependency-path: cdk/yarn.lock

      - name: Build CFN from CDK
        run: ./script/ci
        working-directory: cdk

      # Required by sbt riffRaffUpload
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: ${{ secrets.GU_RIFF_RAFF_ROLE_ARN }}
          aws-region: eu-west-1

      - name: Setup Java 11
        uses: actions/setup-java@v4
        with:
          distribution: "corretto"
          java-version: "11"
      - uses: actions/cache@v3
        with:
          path: |
            ~/.ivy2/cache
            ~/.sbt
            ~/.coursier
          key: sbt

      - name: Build and upload to RiffRaff
        run: |
          export LAST_TEAMCITY_BUILD=20000
          export GITHUB_RUN_NUMBER=$(( $GITHUB_RUN_NUMBER + $LAST_TEAMCITY_BUILD ))
          sbt "project acquisition-events-api" riffRaffUpload
