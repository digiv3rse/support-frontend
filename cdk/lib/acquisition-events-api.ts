import { DomainName, HttpApi } from "@aws-cdk/aws-apigatewayv2-alpha";
import { GuApiLambda } from "@guardian/cdk";
import { GuAlarm } from "@guardian/cdk/lib/constructs/cloudwatch";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import type { App } from "aws-cdk-lib";
import { Duration } from "aws-cdk-lib";
import { CfnBasePathMapping, CfnDomainName } from "aws-cdk-lib/aws-apigateway";
import { CfnIntegration, CfnRoute } from "aws-cdk-lib/aws-apigatewayv2";
import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import { ComparisonOperator, Metric } from "aws-cdk-lib/aws-cloudwatch";
import {
  Effect,
  Policy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { CfnRecordSet } from "aws-cdk-lib/aws-route53";

export interface AcquisitionEventsApiProps extends GuStackProps {
  stack: string;
  stage: string;
  certificateId: string;
  domainName: string;
  hostedZoneName: string;
  hostedZoneId: string;
}

export class AcquisitionEventsApi extends GuStack {
  constructor(scope: App, id: string, props: AcquisitionEventsApiProps) {
    super(scope, id, props);

    const app = "acquisition-events-api";

    const commonEnvironmentVariables = {
      App: app,
      Stack: this.stack,
      Stage: this.stage,
    };

    // ---- API-triggered lambda functions ---- //
    // const acquisitionEventsApiLambda = new GuApiLambda(
    //   this,
    //   "acquisition-events-api-cdk-lambda",
    //   {
    //     description:
    //       "A lambda that Sends in-app acquisitions (subscriptions) to BigQuery",
    //     functionName: `${app}-cdk-${this.stage}`,
    //     fileName: `${app}.jar`,
    //     handler: "com.gu.acquisitionEventsApi.Lambda::handler",
    //     runtime: Runtime.JAVA_8,
    //     memorySize: 1024,
    //     timeout: Duration.seconds(300),
    //     environment: commonEnvironmentVariables,
    //     // Create an alarm
    //     monitoringConfiguration: {
    //       http5xxAlarm: { tolerated5xxPercentage: 5 },
    //       snsTopicName: "conversion-dev",
    //     },
    //     app: "acquisition-events-api",
    //     api: {
    //       id: `${app}-${this.stage}`,
    //       description: "API Gateway created by CDK",
    //     },
    //   }
    // );

    // ---- Alarms ---- //
    const alarmName = (shortDescription: string) =>
      `ACQUISITION-EVENTS-API-CDK- ${this.stage} ${shortDescription}`;

    const alarmDescription = (description: string) =>
      `Impact - ${description}. Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit`;

    new GuAlarm(this, "ApiGateway4XXAlarmCDK", {
      app,
      alarmName: alarmName("API gateway 4XX response"),
      alarmDescription: alarmDescription(
        "Acquisition Events API received an invalid request"
      ),
      evaluationPeriods: 1,
      threshold: 1,
      snsTopicName: "contributions-dev",
      actionsEnabled: this.stage === "PROD",
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      metric: new Metric({
        metricName: "4XXError",
        namespace: "AWS/ApiGateway",
        statistic: "Sum",
        period: Duration.seconds(300),
        dimensionsMap: {
          ApiName: `${app}-${this.stage}`,
        },
      }),
    });

    // ---- DNS ---- //
    const certificateArn = `arn:aws:acm:eu-west-1:${this.account}:certificate/${props.certificateId}`;
    // const cfnDomainName = new CfnDomainName(this, "DomainName", {
    //   domainName: props.domainName,
    //   regionalCertificateArn: certificateArn,
    //   endpointConfiguration: {
    //     types: ["REGIONAL"],
    //   },
    // });
    //
    // new CfnBasePathMapping(this, "BasePathMapping", {
    //   domainName: cfnDomainName.ref,
    //   // Uncomment the lines below to reroute traffic to the new API Gateway instance
    //   restApiId: acquisitionEventsApiLambda.api.restApiId,
    //   stage: acquisitionEventsApiLambda.api.deploymentStage.stageName,
    //   // Uncomment the lines below to reroute traffic to the old (existing) API Gateway instance
    //   // restApiId: yamlDefinedResources.getResource("ServerlessRestApi").ref,
    //   // stage: props.stage,
    // });
    //
    // new CfnRecordSet(this, "DNSRecord", {
    //   name: props.domainName,
    //   type: "CNAME",
    //   hostedZoneId: props.hostedZoneId,
    //   ttl: "120",
    //   resourceRecords: [cfnDomainName.attrRegionalDomainName],
    // });

    // ---- Eventbridge stuff ---- //
    const eventBusArn = `arn:aws:events:eu-west-1:865473395570:event-bus/acquisitions-bus-${props.stage}`;
    // There's no Eventbridge integration available as CDK L2 yet, so we have to use L1 and create Role, Integration and Route
    // https://docs.aws.amazon.com/cdk/api/v2/docs/aws-apigatewayv2-alpha-readme.html#defining-http-apis
    const dn = new DomainName(this, "DomainName", {
      domainName: props.domainName,
      certificate: Certificate.fromCertificateArn(this, "cert", certificateArn),
    });

    const httpApi = new HttpApi(
      this,
      `acquisitions-eventbridge-api-${props.stage}`,
      {
        defaultDomainMapping: {
          domainName: dn,
          mappingKey: "foo",
        },
      }
    );

    // Role
    const apiRole = new Role(this, "EventBridgeIntegrationRole", {
      assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
    });

    apiRole.addToPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: [eventBusArn],
        actions: ["events:PutEvents"],
      })
    );

    // Integration
    const eventBridgeIntegration = new CfnIntegration(
      this,
      "EventBridgeIntegration",
      {
        apiId: httpApi.httpApiId,
        integrationType: "AWS_PROXY",
        integrationSubtype: "EventBridge-PutEvents",
        credentialsArn: apiRole.roleArn,
        requestParameters: {
          Source: "ApiGateway",
          DetailType: "AcquisitionsEvent",
          Detail: "$request.body",
          EventBusName: eventBusArn,
        },
        payloadFormatVersion: "1.0",
        timeoutInMillis: 10000,
      }
    );

    // Route
    new CfnRoute(this, "EventRoute", {
      apiId: httpApi.httpApiId,
      routeKey: "POST /",
      target: `integrations/${eventBridgeIntegration.ref}`,
    });

    // ---- Apply policies ---- //
    const ssmInlinePolicy: Policy = new Policy(this, "SSM inline policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["ssm:GetParametersByPath"],
          resources: [
            `arn:aws:ssm:${this.region}:${this.account}:parameter/acquisition-events-api/bigquery-config/${props.stage}/*`,
          ],
        }),
      ],
    });

    const s3InlinePolicy: Policy = new Policy(this, "S3 inline policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:GetObject"],
          resources: ["arn:aws:s3::*:membership-dist/*"],
        }),
      ],
    });

    // acquisitionEventsApiLambda.role?.attachInlinePolicy(ssmInlinePolicy);
    // acquisitionEventsApiLambda.role?.attachInlinePolicy(s3InlinePolicy);
  }
}
