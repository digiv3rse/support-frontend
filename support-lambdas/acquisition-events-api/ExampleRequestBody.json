{"eventTimeStamp":"2018-12-13T14:15:04.165Z","product":{"RecurringContribution":{}},"amount":1.0,"country":{"alpha2":"GB","name":"United Kingdom","statesByCode":{}},"currency":{"GBP":{}},"componentId":"MY_COMPONENT_ID","componentType":null,"campaignCode":"MY_CAMPAIGN_CODE","source":null,"referrerUrl":"referrer-url","abTests":[],"paymentFrequency":{"Monthly":{}},"paymentProvider":{"Stripe":{}},"printOptions":null,"browserId":null,"identityId":null,"pageViewId":null,"referrerPageViewId":null,"labels":[],"promoCode":null,"reusedExistingPaymentMethod":false,"readerType":"Direct","acquisitionType":{"Purchase":{}},"zuoraSubscriptionNumber":null,"zuoraAccountNumber":null,"contributionId":null,"paymentId":null,"queryParameters":[],"platform":null}
