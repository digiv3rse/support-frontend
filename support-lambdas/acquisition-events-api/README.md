# Acquisition Events Api
- API for sending in-app acquisitions(subscriptions) to BigQuery using a lambda

[Architecture diagram](https://docs.google.com/drawings/d/10H4fWh2byXRIuV82IyEJkQzMnHX6eCto3GLYtHiBrNE/edit)
## Testing
To test locally:
- Comment out `@Ignore` from `LambdaSpec.scala`
- Run: `AWS_PROFILE=membership sbt`
- In sbt, run: `project acquisition-events-api`; `test`

An example json in `ExampleRequestBody.json` to test on AWS API Gateway

## Manually Inserting Failed Events

At the moment (as of 2023-07-28), the acquisition-events-api doesn’t have a dead
letter queue or anything similar, so events can fail to be written to BigQuery.
In these cases, the API can be manually hit with curl to re-send the events.

First, get the event from the [acquisition-events-api logs in
cloudwatch](https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Facquisition-events-api-cdk-PROD).
[This
search](https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Facquisition-events-api-cdk-PROD/log-events$3FfilterPattern$3DFailed+event+was$26start$3D-1800000)
will find the relevant error messages from the last 30 minutes: customise the
time window to find the events you’re looking for. Save the event in a json file
(use individual files per event if you have multiple).

Get the URL for the acquisition-events-api lambda [from the AWS
console](https://eu-west-1.console.aws.amazon.com/lambda/home?region=eu-west-1#/applications/acquisition-events-api-PROD).
The URL you want is the API Endpoint URL, which should end in `/prod`.

Call curl with the endpoint and provide the json file as data:

``` sh
curl -i -X POST "$apiEndpoint" -d @/path/to/acquisition.json
```

(I've included `-i` so you can see the response headers – without them it’s
often not clear whether the request succeeded. As long as you get a 200 it
should be fine.)
