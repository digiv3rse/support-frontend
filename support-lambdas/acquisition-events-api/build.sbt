import LibraryVersions.circeVersion
import LibraryVersions._

name := "acquisition-events-api"
description := "A lambda for acquisitions events api"

libraryDependencies ++= Seq(
  "com.amazonaws" % "aws-lambda-java-core" % "1.2.0",
  "com.amazonaws" % "aws-lambda-java-events" % "3.11.1",
  "com.amazonaws" % "aws-java-sdk-ssm" % awsClientVersion,
  "ch.qos.logback" % "logback-classic" % "1.1.11",
  "com.typesafe.scala-logging" %% "scala-logging" % "3.9.5",
  "io.circe" %% "circe-core" % circeVersion,
  "io.circe" %% "circe-generic" % circeVersion,
)

assemblyJarName := s"${name.value}.jar"
riffRaffPackageType := assembly.value
riffRaffUploadArtifactBucket := Option("riffraff-artifact")
riffRaffUploadManifestBucket := Option("riffraff-builds")
riffRaffManifestProjectName := s"support:lambdas:${name.value}"
riffRaffArtifactResources += (file("cdk/cdk.out/Acquisition-Events-API-PROD.template.json"), "cfn/Acquisition-Events-API-PROD.template.json")
riffRaffArtifactResources += (file("cdk/cdk.out/Acquisition-Events-API-CODE.template.json"), "cfn/Acquisition-Events-API-CODE.template.json")
