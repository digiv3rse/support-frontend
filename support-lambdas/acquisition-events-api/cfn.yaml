---
AWSTemplateFormatVersion: '2010-09-09'
Description: API for acquisition events
Parameters:
  CertificateArn:
    Description: ARN of the certificate
    Type: String
  App:
    Description: Acquisition Events Api
    Type: String
    Default: acquisition-events-api
  DeployBucket:
    Description: Bucket to copy files to
    Type: String
    Default: membership-dist
  Stack:
    Description: Stack name
    Type: String
    Default: support
  Stage:
    Description: Set by RiffRaff on each deploy
    Type: String
    AllowedValues:
      - CODE
      - PROD
Mappings:
  StageMap:
    CODE:
      DomainName: acquisition-events-code.support.guardianapis.com
      CorsOrigin: "'*'"
    PROD:
      DomainName: acquisition-events.support.guardianapis.com
      CorsOrigin: "'*'"
Resources:
  ApiGateway5XXAlarm:
    Type: AWS::CloudWatch::Alarm
    Condition: IsProd
    Properties:
      AlarmActions:
        - Fn::Sub: arn:aws:sns:${AWS::Region}:${AWS::AccountId}:contributions-dev
      AlarmName:
        Fn::Sub: acquisition-events-api-${Stage} API gateway 5XX response
      AlarmDescription: Acquisition events API failed to process an event
      MetricName: 5XXError
      Namespace: AWS/ApiGateway
      Dimensions:
        - Name: ApiName
          Value:
            Fn::Sub: acquisition-events-api-${Stage}
      ComparisonOperator: GreaterThanOrEqualToThreshold
      Threshold: 1
      Period: 60
      EvaluationPeriods: 1
      Statistic: Sum
  DomainName:
    Type: AWS::ApiGateway::DomainName
    Properties:
      RegionalCertificateArn:
        Ref: CertificateArn
      DomainName:
        Fn::FindInMap:
          - StageMap
          - Ref: Stage
          - DomainName
      EndpointConfiguration:
        Types:
          - REGIONAL
  LambdaAcquisitionEventPermissionProd:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Ref: Lambda
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Sub:
          - arn:aws:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/POST/acquisition
          - __ApiId__:
              Ref: ServerlessRestApi
            __Stage__: "*"
  ServerlessRestApiProdStage:
    Type: AWS::ApiGateway::Stage
    Properties:
      DeploymentId:
        Ref: ServerlessRestApiDeployment0eaecd80e2
      RestApiId:
        Ref: ServerlessRestApi
      StageName: Prod
  ServerlessRestApiDeployment0eaecd80e2:
    Type: AWS::ApiGateway::Deployment
    Properties:
      Description: 'RestApi deployment id: 0eaecd80e2ee9fd80faf745412676700de78c061'
      RestApiId:
        Ref: ServerlessRestApi
      StageName: Stage
  LambdaRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Action:
              - sts:AssumeRole
            Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
      Policies:
        - PolicyName: LambdaRolePolicy1
          PolicyDocument:
            Statement:
              Effect: Allow
              Action:
                - ssm:GetParametersByPath
              Resource:
                - Fn::Sub: arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/acquisition-events-api/bigquery-config/${Stage}/*
        - PolicyName: LambdaRolePolicy2
          PolicyDocument:
            Statement:
              - Effect: Allow
                Action: s3:GetObject
                Resource:
                  - arn:aws:s3::*:membership-dist/*
      Tags:
        - Key: lambda:createdBy
          Value: SAM
  ApiGateway4XXAlarm:
    Type: AWS::CloudWatch::Alarm
    Condition: IsProd
    Properties:
      AlarmActions:
        - Fn::Sub: arn:aws:sns:${AWS::Region}:${AWS::AccountId}:contributions-dev
      AlarmName:
        Fn::Sub: acquisition-events-api-${Stage} API gateway 4XX response
      AlarmDescription: Acquisition Events API received an invalid request
      MetricName: 4XXError
      Namespace: AWS/ApiGateway
      Dimensions:
        - Name: ApiName
          Value:
            Fn::Sub: acquisition-events-api-${Stage}
        - Name: Method
          Value: POST
      ComparisonOperator: GreaterThanOrEqualToThreshold
      Threshold: 1
      Period: 300
      EvaluationPeriods: 1
      Statistic: Sum
  BasePathMapping:
    Type: AWS::ApiGateway::BasePathMapping
    Properties:
      RestApiId:
        Ref: ServerlessRestApi
      DomainName:
        Ref: DomainName
      Stage:
        Fn::Sub: Prod
    DependsOn: ServerlessRestApiProdStage
  DNSRecord:
    Type: AWS::Route53::RecordSet
    Properties:
      HostedZoneName: support.guardianapis.com.
      Name:
        Fn::FindInMap:
          - StageMap
          - Ref: Stage
          - DomainName
      Comment:
        Fn::Sub: CNAME for acquisition events endpoints ${Stage}
      Type: CNAME
      TTL: '120'
      ResourceRecords:
        - Fn::GetAtt:
            - DomainName
            - RegionalDomainName
  ServerlessRestApi:
    Type: AWS::ApiGateway::RestApi
    Properties:
      Body:
        info:
          version: '1.0'
          title:
            Ref: AWS::StackName
        paths:
          "/acquisition":
            post:
              x-amazon-apigateway-integration:
                httpMethod: POST
                type: aws_proxy
                uri:
                  Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${Lambda.Arn}/invocations
              responses: {}
            options:
              x-amazon-apigateway-integration:
                type: mock
                requestTemplates:
                  application/json: |
                    {
                      "statusCode" : 200
                    }
                responses:
                  default:
                    statusCode: '200'
                    responseTemplates:
                      application/json: "{}\n"
                    responseParameters:
                      method.response.header.Access-Control-Allow-Origin:
                        Fn::FindInMap:
                          - StageMap
                          - Ref: Stage
                          - CorsOrigin
                      method.response.header.Access-Control-Allow-Methods: "'*'"
                      method.response.header.Access-Control-Allow-Headers: "'Content-Type'"
              consumes:
                - application/json
              summary: CORS support
              responses:
                '200':
                  headers:
                    Access-Control-Allow-Origin:
                      type: string
                    Access-Control-Allow-Headers:
                      type: string
                    Access-Control-Allow-Methods:
                      type: string
                  description: Default response for CORS method
              produces:
                - application/json
        swagger: '2.0'
  Lambda:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        S3Bucket:
          Ref: DeployBucket
        S3Key:
          Fn::Sub: "${Stack}/${Stage}/${App}/${App}.jar"
      Description: A lambda for acquisitions events api
      FunctionName:
        Fn::Sub: "${App}-${Stage}"
      Handler: com.gu.acquisitionEventsApi.Lambda::handler
      MemorySize: 512
      Role:
        Fn::GetAtt:
          - LambdaRole
          - Arn
      Runtime: java8.al2
      Timeout: 300
      Environment:
        Variables:
          STAGE:
            Ref: Stage
      Tags:
        - Key: lambda:createdBy
          Value: SAM
Conditions:
  IsProd:
    Fn::Equals:
      - Ref: Stage
      - PROD

