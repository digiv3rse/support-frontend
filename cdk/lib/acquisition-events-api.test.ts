import "@aws-cdk/assert/jest";
import { App } from "aws-cdk-lib";
import {AcquisitionEventsApi} from "./acquisition-events-api";
import {Template} from "aws-cdk-lib/assertions";

describe("The Acquisition Events API stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new AcquisitionEventsApi(app, "Acquisition-Events-API-PROD", {
      stack: "support",
      stage: "PROD",
    });

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
