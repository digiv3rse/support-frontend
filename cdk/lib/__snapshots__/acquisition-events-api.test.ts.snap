// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Acquisition Events API stack matches the snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": {
    "IsProd": {
      "Fn::Equals": [
        "PROD",
        "PROD",
      ],
    },
  },
  "Description": "API for acquisition events",
  "Mappings": {
    "StageMap": {
      "CODE": {
        "CorsOrigin": "'*'",
        "DomainName": "acquisition-events-code.support.guardianapis.com",
      },
      "PROD": {
        "CorsOrigin": "'*'",
        "DomainName": "acquisition-events.support.guardianapis.com",
      },
    },
  },
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "App": {
      "Default": "acquisition-events-api",
      "Description": "Acquisition Events Api",
      "Type": "String",
    },
    "CertificateArn": {
      "Description": "ARN of the certificate",
      "Type": "String",
    },
    "DeployBucket": {
      "Default": "membership-dist",
      "Description": "Bucket to copy files to",
      "Type": "String",
    },
    "Stack": {
      "Default": "support",
      "Description": "Stack name",
      "Type": "String",
    },
  },
  "Resources": {
    "ApiGateway4XXAlarm": {
      "Condition": "IsProd",
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Sub": "arn:aws:sns:\${AWS::Region}:\${AWS::AccountId}:contributions-dev",
          },
        ],
        "AlarmDescription": "Acquisition Events API received an invalid request",
        "AlarmName": {
          "Fn::Sub": "acquisition-events-api-PROD API gateway 4XX response",
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": {
              "Fn::Sub": "acquisition-events-api-PROD",
            },
          },
          {
            "Name": "Method",
            "Value": "POST",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ApiGateway5XXAlarm": {
      "Condition": "IsProd",
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Sub": "arn:aws:sns:\${AWS::Region}:\${AWS::AccountId}:contributions-dev",
          },
        ],
        "AlarmDescription": "Acquisition events API failed to process an event",
        "AlarmName": {
          "Fn::Sub": "acquisition-events-api-PROD API gateway 5XX response",
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": {
              "Fn::Sub": "acquisition-events-api-PROD",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "BasePathMapping": {
      "DependsOn": [
        "ServerlessRestApiProdStage",
      ],
      "Properties": {
        "DomainName": {
          "Ref": "DomainName",
        },
        "RestApiId": {
          "Ref": "ServerlessRestApi",
        },
        "Stage": {
          "Fn::Sub": "Prod",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "DNSRecord": {
      "Properties": {
        "Comment": {
          "Fn::Sub": "CNAME for acquisition events endpoints PROD",
        },
        "HostedZoneName": "support.guardianapis.com.",
        "Name": {
          "Fn::FindInMap": [
            "StageMap",
            "PROD",
            "DomainName",
          ],
        },
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "DomainName",
              "RegionalDomainName",
            ],
          },
        ],
        "TTL": "120",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "DomainName": {
      "Properties": {
        "DomainName": {
          "Fn::FindInMap": [
            "StageMap",
            "PROD",
            "DomainName",
          ],
        },
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Ref": "CertificateArn",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "Lambda": {
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DeployBucket",
          },
          "S3Key": {
            "Fn::Sub": "\${Stack}/PROD/\${App}/\${App}.jar",
          },
        },
        "Description": "A lambda for acquisitions events api",
        "Environment": {
          "Variables": {
            "STAGE": "PROD",
          },
        },
        "FunctionName": {
          "Fn::Sub": "\${App}-PROD",
        },
        "Handler": "com.gu.acquisitionEventsApi.Lambda::handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "LambdaRole",
            "Arn",
          ],
        },
        "Runtime": "java8.al2",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "lambda:createdBy",
            "Value": "SAM",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaAcquisitionEventPermissionProd": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Ref": "Lambda",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Sub": [
            "arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${__ApiId__}/\${__Stage__}/POST/acquisition",
            {
              "__ApiId__": {
                "Ref": "ServerlessRestApi",
              },
              "__Stage__": "*",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LambdaRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": {
                "Action": [
                  "ssm:GetParametersByPath",
                ],
                "Effect": "Allow",
                "Resource": [
                  {
                    "Fn::Sub": "arn:aws:ssm:\${AWS::Region}:\${AWS::AccountId}:parameter/acquisition-events-api/bigquery-config/PROD/*",
                  },
                ],
              },
            },
            "PolicyName": "LambdaRolePolicy1",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:s3::*:membership-dist/*",
                  ],
                },
              ],
            },
            "PolicyName": "LambdaRolePolicy2",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "lambda:createdBy",
            "Value": "SAM",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerlessRestApi": {
      "Properties": {
        "Body": {
          "info": {
            "title": {
              "Ref": "AWS::StackName",
            },
            "version": "1.0",
          },
          "paths": {
            "/acquisition": {
              "options": {
                "consumes": [
                  "application/json",
                ],
                "produces": [
                  "application/json",
                ],
                "responses": {
                  "200": {
                    "description": "Default response for CORS method",
                    "headers": {
                      "Access-Control-Allow-Headers": {
                        "type": "string",
                      },
                      "Access-Control-Allow-Methods": {
                        "type": "string",
                      },
                      "Access-Control-Allow-Origin": {
                        "type": "string",
                      },
                    },
                  },
                },
                "summary": "CORS support",
                "x-amazon-apigateway-integration": {
                  "requestTemplates": {
                    "application/json": "{
  "statusCode" : 200
}
",
                  },
                  "responses": {
                    "default": {
                      "responseParameters": {
                        "method.response.header.Access-Control-Allow-Headers": "'Content-Type'",
                        "method.response.header.Access-Control-Allow-Methods": "'*'",
                        "method.response.header.Access-Control-Allow-Origin": {
                          "Fn::FindInMap": [
                            "StageMap",
                            "PROD",
                            "CorsOrigin",
                          ],
                        },
                      },
                      "responseTemplates": {
                        "application/json": "{}
",
                      },
                      "statusCode": "200",
                    },
                  },
                  "type": "mock",
                },
              },
              "post": {
                "responses": {},
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${Lambda.Arn}/invocations",
                  },
                },
              },
            },
          },
          "swagger": "2.0",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ServerlessRestApiDeployment0eaecd80e2": {
      "Properties": {
        "Description": "RestApi deployment id: 0eaecd80e2ee9fd80faf745412676700de78c061",
        "RestApiId": {
          "Ref": "ServerlessRestApi",
        },
        "StageName": "Stage",
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ServerlessRestApiProdStage": {
      "Properties": {
        "DeploymentId": {
          "Ref": "ServerlessRestApiDeployment0eaecd80e2",
        },
        "RestApiId": {
          "Ref": "ServerlessRestApi",
        },
        "StageName": "Prod",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
  },
}
`;
